### PR TITLE
inclusiver upper bound on DiscreteUniform fixes #2592

### LINKED ISSUE #2592: Inconsistent inclusive/exclusive upper bound with DiscreteUniform
The `DiscreteUniform` distribution samples with exclusive upper bounds when used through the `Distribution.dist` API, but uses inclusive upper bound when sampling with theano:

```python
import numpy
import theano
import pymc3

# workaround for https://github.com/pymc-devs/pymc3/issues/562
theano.config.compute_test_value = 'warn'

lower = 0
upper = 5
print("specified   : {} to {}".format(lower, upper))

d = pymc3.DiscreteUniform.dist(lower, upper)

samples1 = d.random(size=10000)

with pymc3.Model() as pmodel:
    pmodel.Var("d", d)
    mt = pymc3.sample(draws=10000)
samples2 = mt["d"]

print("dist.random : {} to {}".format(samples1.min(), samples1.max()))
print("multitrace  : {} to {}".format(samples2.min(), samples2.max()))
```

Output:

 ```bash
specified   : 0 to 5
Assigned Metropolis to d
100%|##########################################################################| 10500/10500 [00:01<00:00, 8503.27it/s]
dist.random : 0 to 4
multitrace  : 0 to 5
Press any key to continue . . .
```

### PR DESCRIPTION
**Pull request by michaelosthege**
draw random samples with inclusive upper bound (just like the theano sampling does)
see #2592 for details

### CODE REVIEW

**Comment by twiecki** [comment]
```
_______________ TestScalarParameterSamples.test_discrete_uniform _______________


self = <pymc3.tests.test_distributions_random.TestScalarParameterSamples object at 0x7f75d635c450>


    def test_discrete_uniform(self):

        def ref_rand(size, lower, upper):

            return st.randint.rvs(lower, upper, size=size)

        pymc3_random_discrete(pm.DiscreteUniform, {'lower': -NatSmall, 'upper': NatSmall},

>                             ref_rand=ref_rand)


pymc3/tests/test_distributions_random.py:511: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 


dist = <class 'pymc3.distributions.discrete.DiscreteUniform'>

paramdomains = {'lower': <pymc3.tests.test_distributions.Domain object at 0x7f75d83bc190>, 'upper': <pymc3.tests.test_distributions.Domain object at 0x7f76446d8450>}

valuedomain = <pymc3.tests.test_distributions.Domain object at 0x7f76446d8110>

ref_rand = <function ref_rand at 0x7f75d83b06e0>, size = 100000, alpha = 0.05

fails = 20


    def pymc3_random_discrete(dist, paramdomains,

                              valuedomain=Domain([0]), ref_rand=None,

                              size=100000, alpha=0.05, fails=20):

        model = build_model(dist, valuedomain, paramdomains)

        domains = paramdomains.copy()

        for pt in product(domains, n_samples=100):

            pt = pm.Point(pt, model=model)

            p = alpha

            # Allow Chisq test to fail (i.e., the samples be different)

            # a certain number of times.

            f = fails

            while p <= alpha and f > 0:

                o = model.named_vars['value'].random(size=size, point=pt)

                e = ref_rand(size=size, **pt)

                o = np.atleast_1d(o).flatten()

                e = np.atleast_1d(e).flatten()

                observed = dict(zip(*np.unique(o, return_counts=True)))

                expected = dict(zip(*np.unique(e, return_counts=True)))

                for e in expected.keys():

                    expected[e] = (observed.get(e, 0), expected[e])

                k = np.array([v for v in expected.values()])

                if np.all(k[:, 0] == k[:, 1]):

                    p = 1.

                else:

                    _, p = st.chisquare(k[:, 0], k[:, 1])

                f -= 1

>           assert p > alpha, str(pt)

E           AssertionError: {'upper': array(3), 'lower': array(-3)}

E           assert 0.0 > 0.05


pymc3/tests/test_distributions_random.py:65: AssertionError
```

**Comment by junpenglao** [comment]
should probably change the above test to `return st.randint.rvs(lower, upper+1, size=size)`

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -447,7 +447,8 @@ def __init__(self, lower, upper, *args, **kwargs):
     def _random(self, lower, upper, size=None):
         # This way seems to be the only to deal with lower and upper
         # as array-like.
-        samples = stats.uniform.rvs(lower, upper - lower - np.finfo(float).eps,
+        samples = stats.uniform.rvs(lower,
```
Comment (fonnesbeck) [inline pymc3/distributions/discrete.py]: We really should be using `randint` here, rather than rounding a continuous sample.

Comment (junpenglao) [reply pymc3/distributions/discrete.py]: agree, it should be replaced with `st.randint.rvs(low, high+1)`

**Comment by michaelosthege** [comment]
I changed it to use `randint` as suggested by @fonnesbeck and the test as suggested by @twiecki 
Test are now passing.

**Comment by junpenglao** [comment]
Thanks @michaelosthege !
@fonnesbeck should we merge now or after the release?

**Comment by fonnesbeck** [comment]
I will merge it and then merge master into the RC PR.

**Pull request merged**